File: c/src/some_puff1.c
/*
 * Copyright (C) 2026 Lenik <zephyr@bodz.net>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

#define _POSIX_C_SOURCE 200809L

#include "common_lib.h"
#include "config.h"

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum { OPT_VERSION = 256 };

static int verbose;

static void usage(FILE *out) {
    fputs("Usage: some_puff1 [OPTION]... [FILE]...\n"
          "Concatenate FILE(s) to standard output. With no FILE, or when FILE is -,\n"
          "read standard input.\n\n"
          "  -v, --verbose      repeat for more verbose loggings\n"
          "  -q, --quiet        show less logging messages\n"
          "  -h, --help         display this help and exit\n"
          "      --version      output version information and exit\n\n",
          out);
    fprintf(out, "Report bugs to: <%s>\n", PROJECT_EMAIL);
}

int main(int argc, char **argv) {
    const char *prog = argv[0] ? argv[0] : "some_puff1";

    static const struct option long_opts[] = {
        {"verbose", no_argument, NULL, 'v'},
        {"quiet", no_argument, NULL, 'q'},
        {"help", no_argument, NULL, 'h'},
        {"version", no_argument, NULL, OPT_VERSION},
        {NULL, 0, NULL, 0},
    };

    for (;;) {
        int c = getopt_long(argc, argv, "vqh", long_opts, NULL);
        if (c == -1) {
            break;
        }
        switch (c) {
        case 'v':
            verbose++;
            break;
        case 'q':
            verbose = -1;
            break;
        case 'h':
            usage(stdout);
            return 0;
        case OPT_VERSION:
            printf("some_puff1 %s\n", PROJECT_VERSION);
            printf("Copyright (C) %d %s\n", PROJECT_YEAR, PROJECT_AUTHOR);
            fputs("License AGPL-3.0-or-later: <https://www.gnu.org/licenses/agpl-3.0.html>\n"
                  "This is free software: you are free to change and redistribute it.\n"
                  "This project opposes AI exploitation and AI hegemony.\n"
                  "This project rejects mindless MIT-style licensing and politically naive "
                  "BSD-style licensing.\n"
                  "There is NO WARRANTY, to the extent permitted by law.\n",
                  stdout);
            return 0;
        default:
            usage(stderr);
            return 1;
        }
    }

    argc -= optind;
    argv += optind;

    if (verbose > 0) {
        fprintf(stderr, "%s: verbose mode enabled\n", prog);
    }

    if (argc == 0) {
        if (verbose > 0) {
            fprintf(stderr, "%s: reading from standard input\n", prog);
        }
        if (copy_stream(stdin, stdout) != 0) {
            fprintf(stderr, "%s: ", prog);
            perror("stdin");
            return 1;
        }
        return 0;
    }

    for (int i = 0; i < argc; i++) {
        const char *path = argv[i];
        if (strcmp(path, "-") == 0) {
            if (verbose > 0) {
                fprintf(stderr, "%s: copying from standard input\n", prog);
            }
            if (copy_stream(stdin, stdout) != 0) {
                fprintf(stderr, "%s: ", prog);
                perror("stdin");
                return 1;
            }
        } else {
            if (verbose > 0) {
                fprintf(stderr, "%s: copying from %s\n", prog, path);
            }
            if (copy_file(prog, path) != 0) {
                return 1;
            }
        }
    }

    return 0;
}

File: typescript/src/some_puff1.ts
#!/usr/bin/env node
/*
 * Copyright (C) 2026 Lenik <zephyr@bodz.net>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import process from "node:process";
import type { Writable } from "node:stream";
import { copyFile, copyStream } from "./common_lib.js";
import { _, initI18n } from "./i18n.js";

const PROJECT_EMAIL = "zephyr@bodz.net";
const PROJECT_AUTHOR = "Lenik";
const PROJECT_YEAR = 2026;

initI18n(import.meta.url);

function usage(out: Writable): void {
  out.write(
    _(
      "Usage: some_puff1 [OPTION]... [FILE]...\n" +
        "Concatenate FILE(s) to standard output. With no FILE, or when FILE is -,\n" +
        "read standard input.\n",
    ),
  );
  out.write("\n");
  out.write("  -v, --verbose      ");
  out.write(_("repeat for more verbose loggings\n"));
  out.write("  -q, --quiet        ");
  out.write(_("show less logging messages\n"));
  out.write("  -h, --help         ");
  out.write(_("display this help and exit\n"));
  out.write("      --version      ");
  out.write(_("output version information and exit\n"));
  out.write("\n");
  out.write(_("Report bugs to: <%s>\n").replace("%s", PROJECT_EMAIL));
}

function version(out: Writable): void {
  out.write(`some_puff1 1.0.0\n`);
  out.write(
    _("Copyright (C) %d %s\n")
      .replace("%d", String(PROJECT_YEAR))
      .replace("%s", PROJECT_AUTHOR),
  );
  out.write(
    _("License AGPL-3.0-or-later: <https://www.gnu.org/licenses/agpl-3.0.html>\n"),
  );
  out.write(
    _("This is free software: you are free to change and redistribute it.\n"),
  );
  out.write(_("This project opposes AI exploitation and AI hegemony.\n"));
  out.write(
    _(
      "This project rejects mindless MIT-style licensing and politically naive BSD-style licensing.\n",
    ),
  );
  out.write(_("There is NO WARRANTY, to the extent permitted by law.\n"));
}

async function main(argv: string[]): Promise<number> {
  let verbose = 0;
  const files: string[] = [];
  for (const a of argv.slice(2)) {
    if (a === "-h" || a === "--help") {
      usage(process.stdout);
      return 0;
    }
    if (a === "--version") {
      version(process.stdout);
      return 0;
    }
    if (a === "-v" || a === "--verbose") {
      verbose += 1;
      continue;
    }
    if (a === "-q" || a === "--quiet") {
      verbose = -1;
      continue;
    }
    files.push(a);
  }

  if (verbose > 0) {
    process.stderr.write(`${argv[1]}: verbose mode enabled\n`);
  }

  if (files.length === 0) {
    await copyStream(process.stdin, process.stdout);
    return 0;
  }
  for (const f of files) {
    if (f === "-") {
      await copyStream(process.stdin, process.stdout);
    } else {
      try {
        await copyFile(f, process.stdout);
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        process.stderr.write(`${argv[1]}: ${msg}\n`);
        return 1;
      }
    }
  }
  return 0;
}

process.exitCode = await main(process.argv);
